rootston/cursor.c

```diff
@@ -290,18 +294,128 @@ static void roots_cursor_press_button(struct roots_cursor *cursor,
 	}
 }
 
-void roots_cursor_handle_motion(struct roots_cursor *cursor,
+// TODO: Match on wl_pointer
+static void cursor_constrain(struct roots_cursor *roots_cur, void *data,
+		void (*callback)(void *data, struct wlr_pointer_constraint_v1 *constraint, double sx, double sy, struct wlr_box *box)) {
+	struct wlr_surface *wlr_surface = roots_cur->seat->seat->pointer_state.focused_surface;
+	struct wlr_cursor  *cur         = roots_cur->cursor;
+
+	if (wlr_surface) {
+		struct roots_desktop *desktop = roots_cur->seat->input->server->desktop;
+		struct roots_view *view = NULL;
+		wl_list_for_each(view, &desktop->views, link) {
+			if (view->wlr_surface == wlr_surface) break;
+		}
+		if (!view) return;
+
+		double lx = cur->x, ly = cur->y, sx = -1, sy = -1;
+
+		struct wlr_surface *wlr_surface_of_view;
+		view_at(view, lx, ly, &wlr_surface_of_view, &sx, &sy);
+		assert(wlr_surface_of_view == wlr_surface);
+
+		// used when there is no specific region in the constraints
+		struct wlr_box sbox = {
+			.x      = 0,
+			.y      = 0,
+			.width  = view->width,
+			.height = view->height,
+		};
+
+		struct wlr_pointer_constraint_v1 *constraint;
+		wl_list_for_each(constraint, &view->pointer_constraints, link) {
+			assert(wlr_surface_from_resource(constraint->surface) == wlr_surface);
+			struct wlr_box box = {};
+
+			bool in_region;
+			if (constraint->current.region) {
+				pixman_box32_t pbox;
+				in_region = pixman_region32_contains_point(
+						wlr_region_from_resource(constraint->current.region),
+						sx, sy, &pbox);
+				box = wlr_box_from_pixman_box32(pbox);
+			} else {
+				// is the focused surface really always valid?
+				in_region = true;
+				box = sbox;
+			}
+
+			if (!in_region) {
+				if (view->active_constraint == constraint) {
+					wlr_pointer_constraint_v1_send_unconstrained(view->active_constraint);
+					wl_list_remove(&view->active_constraint_destroy.link);
+					view->active_constraint = NULL;
+				}
+			} else if (view->active_constraint != constraint) {
+				if (view->active_constraint) {
+					wlr_pointer_constraint_v1_send_unconstrained(view->active_constraint);
+					wl_list_remove(&view->active_constraint_destroy.link);
+				}
+
+				wl_signal_add(&constraint->events.destroy, &view->active_constraint_destroy);
+				view->active_constraint = constraint;
+				wlr_pointer_constraint_v1_send_constrained(constraint);
+			}
+
+			if (in_region) {
+				callback(data, constraint, sx, sy, wlr_box_empty(&box) ? &sbox : &box);
+				return;
+			}
+		}
+	}
+}
+
+static void cursor_constrain_warp_absolute(void *data, struct wlr_pointer_constraint_v1 *constraint, double sx, double sy, struct wlr_box *box) {
+	double *deltas = (double*)data;
+	if (constraint->type == WLR_POINTER_CONSTRAINT_V1_LOCKED) {
+		deltas[0] = 0, deltas[1] = 0;
+	} else if (!wlr_box_contains_point(box, sx + deltas[0], sy + deltas[1])) {
+		deltas[0] = 0, deltas[1] = 0;
+	}
+}
```

Comment:
It would be better to have wlr_cursor support taking a constraint - it already supports this via wlr_cursor_map_to_region.